subsys/bluetooth/mesh/gen_onoff_srv.c

```diff
@@ -56,60 +56,93 @@ static void onoff_set(struct bt_mesh_model *model, struct bt_mesh_msg_ctx *ctx,
 		return;
 	}
 
+	int err;
+
 	struct bt_mesh_onoff_srv *srv = model->user_data;
 	struct bt_mesh_onoff_status status = { 0 };
 	struct bt_mesh_model_transition transition;
 	struct bt_mesh_onoff_set set;
 
 	uint8_t on_off = net_buf_simple_pull_u8(buf);
 	uint8_t tid = net_buf_simple_pull_u8(buf);
-
-	if (on_off > 1) {
+	bool new_transaction;
+
+	if (on_off >= 0x02) {
+		/* Generic OnOff states:
+		 *  Value           | Description
+		 * -----------------|-----------------
+		 *  0x00            | Off
+		 *  0x01            | On
+		 *  0x02-0xFF       | Prohibited
+		 */
 		return;
 	}
 
 	set.on_off = on_off;
 
-	if (tid_check_and_update(&srv->prev_transaction, tid, ctx) != 0) {
+	new_transaction =
+		(tid_check_and_update(&srv->prev_transaction, tid, ctx) == 0);
+
+	if (new_transaction) {
+		err = transition_get(srv->model, buf, &transition);
+		if (err) {
+			return;
+		}
+		set.transition = &transition;
+
+		srv->handlers->set(srv, ctx, &set, &status);
+
+		if (IS_ENABLED(CONFIG_BT_MESH_SCENE_SRV)) {
+			bt_mesh_scene_invalidate(&srv->scene);
+		}
+	} else if (ack) {
 		/* If this is the same transaction, we don't need to send it
 		 * to the app, but we still have to respond with a status.
 		 */
 		srv->handlers->get(srv, NULL, &status);
-		goto respond;
-	}
-
-	if (buf->len == 2) {
-		model_transition_buf_pull(buf, &transition);
-	} else {
-		bt_mesh_dtt_srv_transition_get(srv->model, &transition);
-	}
-
-	set.transition = &transition;
-
-	srv->handlers->set(srv, ctx, &set, &status);
-
-	if (IS_ENABLED(CONFIG_BT_MESH_SCENE_SRV)) {
-		bt_mesh_scene_invalidate(&srv->scene);
 	}
 
-	(void)bt_mesh_onoff_srv_pub(srv, NULL, &status);
-
-respond:
 	if (ack) {
+		/* The server responds with an associated Generic OnOff Status
+		 * message to the originator.
+		 */
 		rsp_status(model, ctx, &status);
 	}
+
+	if (new_transaction) {
+		/* The server publishes a Generic OnOff Status message to the group
+		 * address configured as the model's Publish Address.
+		 */
+		(void)bt_mesh_onoff_srv_pub(srv, NULL, &status);
+	}
```

Comment:
We're trying to keep the "happy path" at the innermost indentation level, and do early returns as often as we can. 
I think this would be a bit clearer with the previous structure, but without the `goto`.

Something like this, for instance:

```C
	if (tid_check_and_update(&srv->prev_transaction, tid, ctx) != 0) {
		/* If this is the same transaction, we don't need to send it
		 * to the app, but we still have to respond with a status.
		 */
		if (ack) {
			srv->handlers->get(srv, NULL, &status);
			rsp_status(model, ctx, &status);
		}

		return;
	}

	transition_get(srv->model, buf, &transition);

	set.transition = &transition;

	srv->handlers->set(srv, ctx, &set, &status);

	if (IS_ENABLED(CONFIG_BT_MESH_SCENE_SRV)) {
		bt_mesh_scene_invalidate(&srv->scene);
	}

	if (ack) {
		rsp_status(model, ctx, &status);
	}

	(void)bt_mesh_onoff_srv_pub(srv, NULL, &status);
```

This way, the main code path shows the expected behavior, and we don't need to keep track of the new_transaction or ack state for the entire function

Reply:
I agree with you. It would be clearer. I also considered the same structure.